model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.jpa;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+import javax.persistence.EntityManager;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaDelete;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import org.keycloak.connections.jpa.JpaKeycloakTransaction;
+import static org.keycloak.models.jpa.PaginationUtils.paginateQuery;
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.common.StringKeyConvertor;
+import org.keycloak.models.map.common.StringKeyConvertor.UUIDKey;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.QueryParameters;
+import static org.keycloak.models.map.storage.jpa.JpaMapStorageProviderFactory.CLONER;
+import static org.keycloak.utils.StreamsUtil.closing;
+
+public abstract class JpaMapKeycloakTransaction<E extends JpaRootEntity, MapEntity extends AbstractEntity, M> extends JpaKeycloakTransaction implements MapKeycloakTransaction<MapEntity, M> {
+
+    private final Class<E> entityType;
+
+    @SuppressWarnings("unchecked")
+    public JpaMapKeycloakTransaction(Class<E> entityType, EntityManager em) {
+        super(em);
+        this.entityType = entityType;
+    }
+
+    protected abstract void selectCbConstruct(CriteriaQuery<E> query, CriteriaBuilder cb, Root<E> root);
+    protected abstract void setEntityVersion(JpaRootEntity entity);
```

Comment:
Why is this method abstract? It is already contained in `JpaRootEntity` interface. Is there any reason for not using it directly?

Reply:
It's not known what entity we're dealing with. It could be used directly if we introduce mapping between jpa entity and its supported version. Something like following

```
    private static final Map<Class<? extends JpaRootEntity>, Integer> ENTITY_SUPPORTED_VERSION = new HashMap<>();
    static {
        ENTITY_SUPPORTED_VERSION.put(JpaClientScopeEntity.class,       Constants.SUPPORTED_VERSION_CLIENT_SCOPE);
        ENTITY_SUPPORTED_VERSION.put(JpaClientEntity.class,            Constants.SUPPORTED_VERSION_CLIENT);
        ENTITY_SUPPORTED_VERSION.put(JpaRoleEntity.class,              Constants.SUPPORTED_VERSION_ROLE);
    }
```

Then it could be called like `jpaEntity.setEntityVersion(ENTITY_SUPPORTED_VERSION.get(entityType));`

wdyt?

Reply:
Ah, I realized the reason - it's there to set appropriate entity version depending on the entity type. Sorry for confusion